test/tstDetailsTreeConstruction.cpp

```diff
@@ -148,70 +148,101 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(example_tree_construction, DeviceType,
     sorted_morton_codes(i) = b.to_ulong();
   }
 
-  // reference solution for a recursive traversal from top to bottom
-  // starting from root, visiting first the left child and then the right one
-  std::ostringstream ref;
-  ref << "I0"
-      << "I3"
-      << "I1"
-      << "L0"
-      << "L1"
-      << "I2"
-      << "L2"
-      << "L3"
-      << "I4"
-      << "L4"
-      << "I5"
-      << "I6"
-      << "L5"
-      << "L6"
-      << "L7";
-  std::cout << "ref=" << ref.str() << "\n";
-
   // hierarchy generation
   using ArborX::Box;
   using ArborX::Details::makeLeafNode;
-  using ArborX::Details::Node;
+  using Node = ArborX::Details::NodeWithLeftChildAndRope;
+  using ArborX::Details::ROPE_SENTINEL;
+
   Kokkos::View<Node *, DeviceType> leaf_nodes("Testing::leaf_nodes", n);
   Kokkos::View<Node *, DeviceType> internal_nodes("Testing::internal_nodes",
                                                   n - 1);
   for (int i = 0; i < n; ++i)
-    leaf_nodes(i) = makeLeafNode(i, Box{});
+    leaf_nodes(i) = makeLeafNode(typename Node::Tag{}, i, Box{});
+
+  typename DeviceType::execution_space space{};
+
+  Kokkos::View<Box *, DeviceType> primitives("Testing::primitives", n);
+  Kokkos::View<unsigned int *, DeviceType> permutation_indices(
+      "Testing::indices", n);
+  ArborX::iota(space, permutation_indices);
+
+  ArborX::Details::TreeConstruction::generateHierarchy(
+      space, primitives, permutation_indices, sorted_morton_codes, leaf_nodes,
+      internal_nodes);
+
+  Node const *root = internal_nodes.data();
+
   auto getNodePtr = [&leaf_nodes, &internal_nodes](int i) {
     return i < n - 1 ? &internal_nodes(i) : &leaf_nodes(i - n + 1);
   };
+  auto getLeftChild = [&getNodePtr](Node const *node) {
+    return getNodePtr(node->left_child);
+  };
+  auto getRightChild = [&getNodePtr](Node const *node) {
+    return getNodePtr(getNodePtr(node->left_child)->rope);
```

Comment:
Consider having `template <class Node> int getRightChild(Node const*)` in node header since it is also used in tree traversal.

Reply:
I don't think there is a good way to do it. For a node with a rope, you need bvh to access the right child. I could consider adding
```
template<typename Node, typename Tree>
int getRightChild(Node const* node, Tree const* tree);
```
but it looks quite cumbersome. What do you think?

Reply:
Agree.  This is good enough.